New Driver: Vultr

janeczku opened the pull request:
Vultr driver updated to the new driver specs. Replaces PR #538 by @JamesClonk.     
    
* Conforms with new driver specs
* Added flags for ipv6, automatic backups and private networking
* Added pre-create validations for regionID, planID and API credentials
* Added bats integration tests
* Added unit tests
* Added driver documentation
* Added client request throttling to avoid hitting API rate-limit
    
![bats-test-result](https://cloud.githubusercontent.com/assets/3813921/7004288/737e35c0-dc69-11e4-97c2-1d7d7b3ec9e8.png)
    
Signed-off-by: Jan Broer <janeczku@yahoo.de>

hairyhenderson commented:
@janeczku - I'd like to try this out, can you rebase it please? :)

janeczku commented:
@hairyhenderson Can do - probably tomorrow.

janeczku commented:
Rebased the PR. Please test to break :) @hairyhenderson

hairyhenderson commented:
Thanks @janeczku!

hairyhenderson commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,362 @@
+	MachineID         string
+	MachineName       string
+	PublicIP          string
+	PrivateIP         string
```
Can you add an `IPAddress` property too? It should be the same as `PublicIP`, or `PrivateIP` if `PrivateNetworking` is `true`. (see #1043)

hairyhenderson replied:
I misunderstood `PrivateNetworking` :)... So, can you _either_ add `IPAddress` which is in sync with `PublicIP`, or can you rename `PublicIP` to `IPAddress`?

janeczku replied:
I think its best to just rename `PublicIP`. `IPAdress` is expected to be public anyway.

hairyhenderson replied:
:+1:

hairyhenderson commented:
Ok, so here's an issue, but I don't know if it's in this code or if has to do with some of the recent SSH changes:

```
$ machine create -d vultr --vultr-api-key XXX dm-vultr
Creating SSH key...
Creating Vultr virtual machine...
Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [publickey none], no supported methods remain
Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
... (repeat a billion times)
To see how to connect Docker to this machine, run: machine env dm-vultr
$
```

It _appears_ that the resulting VM is working fine, so it might be nothing.

/cc @nathanleclaire @ehazlett @sthulb - I vaguely recall something recently that might be related?

hairyhenderson commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,362 @@
+	}
+
+	if d.PrivateIP == "" || d.PrivateIP == "0" {
+		d.PrivateIP = "N/A"
```
Any reason this is `N/A` instead of empty string? The convention in other drivers is that it gets set to `""`

janeczku replied:
yes indeed, it should be an empty string if the IP is not set.

ehazlett replied:
@hairyhenderson i've seen this when you use an invalid username.

hairyhenderson commented:
@janeczku - looks like the start command is asynchronous. Can you change `Start()` (and probably `Stop()` too) so that it waits until the machine is reachable before returning? I _think_ most other drivers behave this way.

```
$ machine -D stop dm-vultr
command=stop machine=dm-vultr
$ machine -D start dm-vultr
command=start machine=dm-vultr
$ machine ls
NAME         ACTIVE   DRIVER         STATE     URL                        SWARM
dm-vultr     *        vultr          Running   tcp://45.63.7.71:2376      
$ machine -D ssh dm-vultr
dial tcp 45.63.7.71:22: connection refused
$ machine -D ssh dm-vultr
dial tcp 45.63.7.71:22: connection refused
...
$ machine -D ssh dm-vultr
Welcome to Ubuntu 14.04.2 LTS (GNU/Linux 3.13.0-52-generic x86_64)
...
```

hairyhenderson commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,362 @@
+		cli.BoolFlag{
+			EnvVar: "VULTR_IPV6",
+			Name:   "vultr-ipv6",
+			Usage:  "enable ipv6 for droplet",
```
Maybe:
```
Usage: "Enable IPv6 for VPS",
```

janeczku replied:
Done.

hairyhenderson commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,362 @@
+		cli.BoolFlag{
+			EnvVar: "VULTR_PRIVATE_NETWORKING",
+			Name:   "vultr-private-networking",
+			Usage:  "enable private networking for this virtual machine",
```
Should probably say "VPS" instead of "virtual machine"

janeczku replied:
OK

hairyhenderson commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,362 @@
+		cli.BoolFlag{
+			EnvVar: "VULTR_BACKUPS",
+			Name:   "vultr-backups",
+			Usage:  "enable automatic backups for this virtual machine",
```
Should probably say "VPS" instead of "virtual machine"

hairyhenderson commented:
@janeczku - overall, this is looking pretty good. Might be nice too to add a bit more info-level logging, especially when long operations (during `create` for example) are happening.

janeczku commented:
@hairyhenderson thanks for your valuable observations!

> looks like the start command is asynchronous. Can you change Start() (and probably Stop() too) so that it waits until the machine is reachable before returning? I think most other drivers behave this way.

They should :) But at least the DO driver respectively its API package has asynchronous start/stop commands as well. I will implement this in the Vultr driver and since they share most code we can then easily add this change to the DO driver as well.

>Ok, so here's an issue, but I don't know if it's in this code or if has to do with some of the recent SSH changes: Error dialing TCP: ssh: handshake failed ....

I have investigated this. It's not really the driver's problem. This error occurs when the SSHKey has not yet been provisioned to the new instance (but the SSHd is already running). It takes up to 4 minutes until the Vultr instance is fully provisioned. One basically just needs to bump up the `utils.WaitFor()` timeout a minute:

```Go
func WaitFor(f func() bool) error {
	return WaitForSpecific(f, 90, 3*time.Second)
}
```

janeczku commented:
Merged reviewers suggestions, rebased again and squashed.

hairyhenderson commented:
@janeczku looks like you'll need to rebase _again_

janeczku commented:
Hey @hairyhenderson, would you take a final look at the patch and (hopefully) give your OK? I rebased again and added a tiny bit of code making `RancherOS` the default OS for new instances.

/cc @ibuildthecloud @imikushin
[Test builds](https://github.com/janeczku/machine/releases/tag/v0.3.0-vultr) / [Free trial for VULTR](https://www.vultr.com/freetrial/)

janeczku commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,517 @@
+	}
+	return fmt.Errorf("Plan ID %d not available in this region. Available plans: %v", d.PlanID, plans)
+}
+
```
@hairyhenderson FYI, the code below has been added since you last took a look (to allow booting RancherOS).

janeczku commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,517 @@
+	}
+	return nil
+}
+
```
@hairyhenderson FYI, these two functions have been added after you last took a look (to allow booting RancherOS).

hairyhenderson commented:
> Hey @hairyhenderson, would you take a final look at the patch and (hopefully) give your OK? I rebased again and added a tiny bit of code making RancherOS the default OS for new instances.

Yep! I'll have a look. Might need to wait until Monday or Tuesday, but I'll get to this :)

> Test builds / Free trial for VULTR

Thanks for the build! And, too bad the free credit is only for new users ;)

janeczku commented:
@hairyhenderson Thank you!

imikushin commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,517 @@
+// RancherOS - Create iPXE boot script
+func (d *Driver) createBootScript() error {
+	content := `#!ipxe
+set base-url https://releases.rancher.com/os/latest
```
Would `https:` work w/ iPXE?

janeczku replied:
@imikushin Not in every environment. HTTPS support has to be specifically enabled when compiling the iPXE firmware. So for Vultr HTTPS works, for other iPXE environments one might need to use plain HTTP.

janeczku commented:
ping @hairyhenderson  :smirk:

hairyhenderson commented:
@janeczku - hey, sorry, I've been super-busy with a new job and a new baby (born yesterday!) so it might be a little while before I can take a good look at this... In the meantime, looks like there's some merge conflicts again? I'll try to take a look in the next few days :)

janeczku commented:
> hey, sorry, I've been super-busy with a new job and a new baby (born yesterday!)
@hairyhenderson, thanks for the feedback and congrats!

hairyhenderson commented on drivers/vultr/vultr.go:
```diff
@@ -0,0 +1,517 @@
+			EnvVar: "VULTR_OS",
+			Name:   "vultr-os-id",
+			Usage:  "Vultr operating system ID. Default: 159 (RancherOS)",
+			Value:  159,
```
So... While RancherOS is definitely cool, and is considered "supported" from docker-machine's point of view, the default for remote drivers (like Vultr) is Ubuntu (see https://github.com/docker/machine/blob/master/docs/index.md#specify-a-base-operating-systems). So IMO you should change the default to Ubuntu (probably 14.04 - i.e. `160`), just to keep in line with what everyone expects and assumes.

imikushin replied:
Not that I disagree, but...
RancherOS can be configured to look like Ubuntu :)
It's just a matter of what console container to use. Currently, RancherOS
does use ubuntu-console by default in cloud deployments.

Just my 2 cents.

On Tue, Jun 23, 2015, 23:41 Dave Henderson <notifications@github.com> wrote:

> In drivers/vultr/vultr.go
> <https://github.com/docker/machine/pull/958#discussion_r33076225>:
>
> > +	})
> > +}
> > +
> > +// GetCreateFlags registers the flags this driver adds to "docker hosts create"
> > +func GetCreateFlags() []cli.Flag {
> > +	return []cli.Flag{
> > +		cli.StringFlag{
> > +			EnvVar: "VULTR_API_KEY",
> > +			Name:   "vultr-api-key",
> > +			Usage:  "Vultr API key",
> > +		},
> > +		cli.IntFlag{
> > +			EnvVar: "VULTR_OS",
> > +			Name:   "vultr-os-id",
> > +			Usage:  "Vultr operating system ID. Default: 159 (RancherOS)",
> > +			Value:  159,
>
> So... While RancherOS is definitely cool, and is considered "supported"
> from docker-machine's point of view, the default for remote drivers (like
> Vultr) is Ubuntu (see
> https://github.com/docker/machine/blob/master/docs/index.md#specify-a-base-operating-systems).
> So IMO you should change the default to Ubuntu (probably 14.04 - i.e. 160),
> just to keep in line with what everyone expects and assumes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/pull/958/files#r33076225>.
>

hairyhenderson replied:
Yup, sure. But even though it might _look_ like Ubuntu, it's not Ubuntu ;) I really think all of the remote drivers need to have the same default image.

janeczku replied:
@hairyhenderson I have changed the default OS back to Ubuntu. Since the iPXE deployment of RancherOS is the better choice for Vultr (deploying Ubuntu takes up to 5 minutes) i have added instructions to the docs concerning the possibility to use RancherOS as base OS.

/cc @imikushin

hairyhenderson commented:
@janeczku - so far this looks pretty good! Just the one thing about the default for `--vultr-os-id`, then IMO this is ready to merge :)

/cc @ehazlett @nathanleclaire

janeczku commented:
@ehazlett @nathanleclaire you wanna chime in?  i'd be happy to finally see this merged :)

djmaze commented:
Thanks for your work @janeczku! I just tried out this fork with my Vultr account. It does not matter if I choose RancherOS or Ubuntu, `docker-machine create` always fails like this:

```
Validating Vultr VPS parameters...
Creating Vultr VPS...
Waiting for IP address to become available...
Error creating machine: json: cannot unmarshal string into Go value of type float64
You will want to check the provider to make sure the machine and associated resources were properly removed.
```

Looking at my Vultr console, the servers have been created successfully. Also, I am able to login using the SSH keys created by docker-machine (in `~/.docker/machine/...`). 

Any idea what's going wrong here? How to debug this?

janeczku commented:
@djmaze thanks for your feedback. This is caused by recent changes to Vultr's API protocol. It is already fixed in the Vultr API client and i will push it to the machine driver later today.

janeczku commented:
@djmaze this is now fixed. You can try the dev builds available here: https://github.com/janeczku/machine/releases/tag/v0.4.0-dev
Please report back if you find any other issues!

djmaze commented:
@janeczku Great, it works now. Thanks for the update!

djmaze commented:
After creating multiple machines on Vulr, `docker-machine ls` gives the following:

```
error getting state for host docker3: Rate limit reached - please try your request again later.  Current rate limit: 1 request/sec
```

It seems rate limiting does not work as needed.

janeczku commented:
@djmaze Thanks for your feedback. The rate limit error is due to the `ls` command querying the state of the hosts asynchronously (dispatching a goroutine for every host). The driver has no way to throttle these concurrent queries. This has to be addressed upstream in the libmachine code: https://github.com/docker/machine/blob/master/libmachine/host.go#L381

djmaze commented:
Excuse my limited Go knowledge. The tokenbucket library uses a mutex. Shouldn't this work across goroutines already?

Ah I see, there is one driver instance per machine, so one bucket per host. That's unfortunate ;/

janeczku commented:
@djmaze You got that right. Vultr really needs to change their 1/sec rate limit policy to something more reasonable like x/minute or x/hour.

djmaze commented:
So forgive me if this is stupid, but I just made the token bucket global:

```diff
diff --git a/drivers/vultr/vultr.go b/drivers/vultr/vultr.go
index 11249b7..98a3b04 100644
--- a/drivers/vultr/vultr.go
+++ b/drivers/vultr/vultr.go
@@ -96,10 +96,12 @@ func GetCreateFlags() []cli.Flag {
        }
 }
 
+var vultrTokenBucket = tokenbucket.NewBucket(1*time.Second, 1)
+
 func NewDriver(machineName string, storePath string, caCert string, privateKey string) (drivers.Driver, error) {
        d := &Driver{MachineName: machineName, storePath: storePath, CaCertPath: caCert, PrivateKeyPath: privateKey, bucket: nil}
        // throttle API queries according to 1/s rate-limit
-       d.bucket = tokenbucket.NewBucket(1*time.Second, 1)
+       d.bucket = vultrTokenBucket
        return d, nil
 }
```

This makes the `ls` command work for me. The tockenbucket code itself should be safe from parallel use in goroutines because it uses mutex and channels, right? 

Of course we can do better and initialize the bucket on first use instead (using a mutex). Apart from that, is there anything wrong with this change?

janeczku commented:
Thats interesting! I would not have expected a single tokenbucket to work across multiple goroutines. It was my believe that when you have multiple receivers (`<-d.bucket.SpendToken(1)`) on the same channel,  they would all unblock once there is a single transmit on that channel and therefore breaking the throttling mechanism. It turns out thats not the case (with Golang channels there is only one active receiver at a time, the other ones are "waiting in line").
Again, great find :+1:

janeczku commented:
Confirmed working 1req/s API query throttling.

> Janek$ ./docker-machine_darwin-amd64 -D ls
GetState() timestamp: 2015-07-21 00:45:**38**.064903604 +0200 CEST
GetState() timestamp: 2015-07-21 00:45:**39**.06484465 +0200 CEST
GetState() timestamp: 2015-07-21 00:45:**40**.064840547 +0200 CEST
GetState() timestamp: 2015-07-21 00:45:**41**.06476442 +0200 CEST

djmaze commented:
Unfortunately, now I am not able to build machines based on RancherOS with the Vultr driver. This is the output when using `--vulr-os-id=159`:

```
Validating Vultr VPS parameters...
Creating Vultr VPS...
Deploying RancherOS
Error creating machine: json: cannot unmarshal number into Go value of type string
```
With the default Ubuntu OS the installation works.

djmaze commented:
Got it working again by making [SCRIPTID](https://github.com/JamesClonk/vultr/blob/master/lib/scripts.go#L7) an `int` and adjusting the rest of the code accordingly. So again, this has to be fixed upstream first.

janeczku commented:
@djmaze Pushed fix to the Vultr API client repo. It was not enough to change the SCRIPTID field to integer since the API continues to return this field as a string in other calls. See here: https://github.com/JamesClonk/vultr/pull/7/files

ehazlett commented:
Hi , thanks for your efforts and persistence in submitting this driver. We are extremely excited that there is so much interest in Docker Machine and we really appreciate your interest. However, at this time it is proving to be extremely difficult for us to keep up with reviewing and testing each of these drivers for inclusion in the Machine core. We really want to switch to a more pluggable model, as well as polish up a few things about the driver model which need to be changed to ensure a smooth and sustainable future.

Therefore, we will be moving to a plugin model for 0.5 and would love to have you involved in the design and development process. We are closing the outstanding driver PRs at this time, but please keep the code. We will stick closely to the current driver interface and you should be able to re-use a lot (if not all) of the existing driver along with the new plugin model. We will be moving all of the drivers which are merged directly into Machine today to the plugin model when it is available, so there will be no special treatment of those, and there will be documentation outlining the process of developing and using a Docker Machine driver plugin.

With all of that being said, we want to apologize for the lack of feedback on your pull request. As contributors ourselves, we understand that being left in limbo is no fun. We would have liked to address this sooner, and in the future we will be more responsive around these kinds of issues.

Once again, we thank you for the contribution and the tremendous support. Keep hacking strong!

If you want to contribute to the design of the plugin model, we'd love to get your input on this issue where we will be planning it:

https://github.com/docker/machine/issues/1626

@ehazlett & @nathanleclaire
